### PR TITLE
[feat] Move to cgranges from intervaltree for OverlapDetector

### PR DESCRIPTION
**Pull request by nh13**
This results in speedups >50x when there are (a) >10K intervals and (b) all the intervals are added before querying.  Results in other use cases also show a 1.5-2.5x speedup (e.g. smaller # of intervals or iteratively adding then querying).

### Testing 

To test the speed differences, I did some very basic, non-comprehensive simulations to verify the performance claimed on the [cgranges readme](https://github.com/lh3/cgranges#test-with-bed-coverage).

I varied three variables:

1. `total-intervals` (X): I randomly simulated `X` intervals from a single contig of length 100M where each interval was 100bp in length.
2. `target-intervals` (Y): I randomly picked `Y < X` intervals from the total-intervals simulated in (1) to _add to_ the overlap detector.
3. `query-intervals` (Z): I randomly picked `Z` intervals from the total-intervals simulated in (1) to _query_ the overlap detector.

I then varied `X`, `Y`, and `Z` across three different access patterns:

1. `add_all`: I add all `Y` intervals to the detector, then query each of the `Z` intervals one-by-one.  This tests the recommended or expected usage pattern: load in a bunch of intervals, then query.  Varying `X/Y/Z` shows both scale and fraction of misses.
2. `no_targets`: the detector is empty, then query each of the `Z` intervals one-by-one.  This tests when the detector is empty and misses 100% of the time.
3. `one_by_one`: iteratively add an _query_ interval and then query for it.  This tests how expensive it is to index the interval tree, which `cgranges` does and `intervaltree` doesn't (or at least doesn't expose).

The results below show a speedup in all cases.  This gets significnat when the # of intervals is large (>10K) and the # of queries is large.

<details>

<summary>Speed testing between cgranges and intervaltree</summary>

```
test        total-intervals  target-intervals  query-intervals  speedup  intervaltree  cgranges
add_all     1000             1000              1000             24.44    0.02          0.00
add_all     1000             1000              10000            26.31    0.15          0.01
add_all     1000             1000              100000           25.62    1.48          0.06
add_all     1000             1000              1000000          25.46    14.69         0.58
add_all     10000            1000              1000             72.10    0.05          0.00
add_all     10000            1000              10000            82.94    0.47          0.01
add_all     10000            1000              100000           84.32    4.72          0.06
add_all     10000            1000              1000000          88.83    49.95         0.56
add_all     10000            10000             1000             105.46   0.11          0.00
add_all     10000            10000             10000            168.54   1.18          0.01
add_all     10000            10000             100000           182.24   11.89         0.07
add_all     10000            10000             1000000          185.05   118.46        0.64
add_all     100000           1000              1000             84.12    0.05          0.00
add_all     100000           1000              10000            83.83    0.55          0.01
add_all     100000           1000              100000           79.43    5.40          0.07
add_all     100000           1000              1000000          92.28    54.20         0.59
add_all     100000           10000             1000             410.86   0.44          0.00
add_all     100000           10000             10000            594.53   4.42          0.01
add_all     100000           10000             100000           667.37   45.52         0.07
add_all     100000           10000             1000000          645.00   429.96        0.67
add_all     100000           100000            1000             162.78   0.96          0.01
add_all     100000           100000            10000            822.48   10.74         0.01
add_all     100000           100000            100000           1362.21  110.54        0.08
add_all     100000           100000            1000000          1494.81  1067.96       0.71
add_all     1000000          1000              1000             89.18    0.06          0.00
add_all     1000000          1000              10000            91.47    0.54          0.01
add_all     1000000          1000              100000           81.98    5.36          0.07
add_all     1000000          1000              1000000          80.90    54.45         0.67
add_all     1000000          10000             1000             469.45   0.51          0.00
add_all     1000000          10000             10000            731.36   5.13          0.01
add_all     1000000          10000             100000           747.58   52.63         0.07
add_all     1000000          10000             1000000          719.37   509.38        0.71
add_all     1000000          100000            1000             854.44   4.80          0.01
add_all     1000000          100000            10000            3797.51  48.04         0.01
add_all     1000000          100000            100000           6137.77  520.82        0.08
add_all     1000000          100000            1000000          6575.94  5212.83       0.79
add_all     1000000          1000000           1000             134.04   6.13          0.05
add_all     1000000          1000000           10000            1088.68  59.26         0.05
add_all     1000000          1000000           100000           4135.11  597.58        0.14
add_all     1000000          1000000           1000000          6420.71  6497.30       1.01
no_targets  1000             1000              1000             1.20     0.00          0.00
no_targets  1000             1000              10000            1.24     0.00          0.00
no_targets  1000             1000              100000           1.25     0.02          0.02
no_targets  1000             1000              1000000          1.22     0.19          0.16
no_targets  10000            1000              1000             1.18     0.00          0.00
no_targets  10000            1000              10000            1.23     0.00          0.00
no_targets  10000            1000              100000           1.24     0.02          0.02
no_targets  10000            1000              1000000          1.25     0.20          0.16
no_targets  10000            10000             1000             1.19     0.00          0.00
no_targets  10000            10000             10000            1.21     0.00          0.00
no_targets  10000            10000             100000           1.23     0.02          0.02
no_targets  10000            10000             1000000          1.24     0.19          0.16
no_targets  100000           1000              1000             1.16     0.00          0.00
no_targets  100000           1000              10000            1.35     0.00          0.00
no_targets  100000           1000              100000           1.34     0.03          0.02
no_targets  100000           1000              1000000          1.25     0.23          0.18
no_targets  100000           10000             1000             1.21     0.00          0.00
no_targets  100000           10000             10000            1.32     0.00          0.00
no_targets  100000           10000             100000           1.28     0.02          0.02
no_targets  100000           10000             1000000          1.23     0.23          0.19
no_targets  100000           100000            1000             1.33     0.00          0.00
no_targets  100000           100000            10000            1.37     0.00          0.00
no_targets  100000           100000            100000           1.32     0.03          0.02
no_targets  100000           100000            1000000          1.24     0.21          0.17
no_targets  1000000          1000              1000             1.27     0.00          0.00
no_targets  1000000          1000              10000            1.29     0.00          0.00
no_targets  1000000          1000              100000           1.33     0.03          0.02
no_targets  1000000          1000              1000000          1.24     0.31          0.25
no_targets  1000000          10000             1000             1.27     0.00          0.00
no_targets  1000000          10000             10000            1.35     0.00          0.00
no_targets  1000000          10000             100000           1.36     0.03          0.02
no_targets  1000000          10000             1000000          1.14     0.31          0.27
no_targets  1000000          100000            1000             1.50     0.00          0.00
no_targets  1000000          100000            10000            1.40     0.00          0.00
no_targets  1000000          100000            100000           1.35     0.03          0.02
no_targets  1000000          100000            1000000          1.23     0.31          0.25
no_targets  1000000          1000000           1000             1.84     0.00          0.00
no_targets  1000000          1000000           10000            2.02     0.00          0.00
no_targets  1000000          1000000           100000           1.64     0.03          0.02
no_targets  1000000          1000000           1000000          1.14     0.31          0.27
one_by_one  1000             1000              1000             5.91     0.00          0.00
one_by_one  1000             1000              10000            6.50     0.03          0.00
one_by_one  1000             1000              100000           6.39     0.31          0.05
one_by_one  1000             1000              1000000          6.48     3.11          0.48
one_by_one  10000            1000              1000             6.88     0.00          0.00
one_by_one  10000            1000              10000            8.41     0.04          0.00
one_by_one  10000            1000              100000           8.54     0.41          0.05
one_by_one  10000            1000              1000000          8.64     4.17          0.48
one_by_one  10000            10000             1000             6.95     0.00          0.00
one_by_one  10000            10000             10000            8.48     0.04          0.00
one_by_one  10000            10000             100000           8.47     0.41          0.05
one_by_one  10000            10000             1000000          8.48     4.10          0.48
one_by_one  100000           1000              1000             6.86     0.00          0.00
one_by_one  100000           1000              10000            8.98     0.04          0.00
one_by_one  100000           1000              100000           13.31    0.65          0.05
one_by_one  100000           1000              1000000          11.89    5.78          0.49
one_by_one  100000           10000             1000             6.68     0.00          0.00
one_by_one  100000           10000             10000            8.65     0.04          0.01
one_by_one  100000           10000             100000           11.09    0.54          0.05
one_by_one  100000           10000             1000000          11.79    5.78          0.49
one_by_one  100000           100000            1000             7.23     0.00          0.00
one_by_one  100000           100000            10000            8.92     0.04          0.00
one_by_one  100000           100000            100000           12.23    0.59          0.05
one_by_one  100000           100000            1000000          11.78    5.57          0.47
one_by_one  1000000          1000              1000             6.82     0.00          0.00
one_by_one  1000000          1000              10000            8.94     0.04          0.00
one_by_one  1000000          1000              100000           13.95    0.67          0.05
one_by_one  1000000          1000              1000000          17.23    8.69          0.50
one_by_one  1000000          10000             1000             6.88     0.00          0.00
one_by_one  1000000          10000             10000            9.13     0.04          0.00
one_by_one  1000000          10000             100000           11.45    0.55          0.05
one_by_one  1000000          10000             1000000          17.76    8.61          0.48
one_by_one  1000000          100000            1000             6.97     0.00          0.00
one_by_one  1000000          100000            10000            9.50     0.04          0.00
one_by_one  1000000          100000            100000           14.43    0.70          0.05
one_by_one  1000000          100000            1000000          16.82    8.04          0.48
one_by_one  1000000          1000000           1000             7.09     0.00          0.00
one_by_one  1000000          1000000           10000            8.79     0.04          0.00
one_by_one  1000000          1000000           100000           11.47    0.55          0.05
one_by_one  1000000          1000000           1000000          17.53    8.37          0.48
```

</details>

### CODE REVIEW

File: pybedlite/overlap_detector.py [hunk 1]
```diff
@@ -100,26 +101,47 @@ def length(self) -> int:
...
+
+    This detector is the most efficient when all intervals are added ahead of time.
     """
 
     def __init__(self) -> None:
-        self._refname_to_tree: Dict[str, IntervalTree] = {}
+        # A mapping from the contig/chromosome name to the associated interval tree
+        self._refname_to_tree: Dict[str, cr.cgranges] = {}
+        self._refname_to_indexed: Dict[str, bool] = {}
+        self._refname_to_intervals: Dict[str, List[Interval]] = {}
```
Comment (nh13) [inline pybedlite/overlap_detector.py:118]: I could be convinced making this a a single dict to a dataclass, but it seemed heavy weight when these are only used in the `__init__` and `add` methods only.

Comment (NatPRoach) [reply pybedlite/overlap_detector.py:118]: Hm, that seems reasonable to me.

File: pybedlite/overlap_detector.py [hunk 2]
```diff
@@ -100,26 +101,47 @@ def length(self) -> int:
...
+        super().__init__()
```
Comment (NatPRoach) [inline pybedlite/overlap_detector.py:119]: What is this class inheriting from that it's calling the `super()` init? On my reading it looks like it's just inheriting from object?

**Review by NatPRoach — APPROVED** [review]
This LGTM, thanks @nh13 !

File: pybedlite/overlap_detector.py (hunk 2 above)
Comment (nh13) [reply pybedlite/overlap_detector.py:119]: Ooops holdover from the speed testing code, thank-you!!

```suggestion
```

**Comment by clintval** [comment]
@nh13 do you still have your test files lying around?

**Comment by nh13** [comment]
@clintval I looked quickly and couldn't find it, so I'll have to dig on my old laptop.

**Pull request merged**